### PR TITLE
[AUTH-1481] [AUTH-1492] Make tests compatible with Remote selenium; DRY some things out for maintainability

### PR DESCRIPTION
**Pull request by tomthorogood**
# Change Summary

When reading the actual code changes, you may want to **go commit-by-commit, and _skip_ b82447c** (the 'webdriver-recorder 4.0' commit). That commit is huge, but mostly repeated changes, which are explained in detail below (to save you from having to read the monstrosity to figure it out). 

[Storyboards for this change (manually uploaded)](https://identity-artifact.iamdev.s.uw.edu/idp-web-tests/sha/b82447c.html)

---

## Why the big diff?

Most of the changes can be summed up with the following examples:

### Example 1: Logging in to the idp

This:

```python
#  Happy case
browser.get(f'{sp_url(sp)}/shib{self.test_env}mfa')
browser.wait_for_tag('p', 'Please sign in')
browser.send_inputs(netid3, password)
browser.click(Locators.submit_button)
browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```

was reduced to this:

```python
browser.get(sp_shib_url(sp, append='mfa'))
log_in_netid(browser, netid3, match_service_provider=sp)
```
### Example 2: Entering the duo passcode

Before:

```python
browser.execute_script("document.getElementsByClassName('passcode-input')[0].value=arguments[0]", passcode)
browser.snap()
browser.execute_script("document.getElementById('passcode').click()")
browser.wait_for_tag('span', 'Incorrect passcode. Enter a passcode from Duo Mobile.')
browser.execute_script("document.getElementById('passcode').click()")
browser.execute_script("document.getElementsByClassName('passcode-input')[0].value=arguments[0]", self.passcode)
browser.snap()
browser.execute_script("document.getElementById('passcode').click()")
browser.switch_to.default_content()
browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```

After:

```python
enter_duo_passcode(browser, match_service_provider=sp)   # Happy case
```

### 

Since those two flows are repeated dozens of times each, you can see how factoring them resulted in such a large, but honestly pretty boring, diff. The examples only show happy cases; the refactor accounts for all variations on this. Keep reading for more information and links to the _important_ bits of code.

---

# Developer Impact

## AWS Credentials

AWS Credentials are no longer required setup for developers. Instead, those credentials will be 
accessed using the GOOGLE_APPLICATION_CREDENTIALS that are already required for 
other packages.

## Setting up and running tests

In order to prepare to use the shared poetry docker image,
I migrated this package to use poetry, which simplifies the setup process 
and makes it easier for everyone, including docker images, to set up and 
run these tests.

After this change, developers should follow the steps in README.md to ensure they are properly set up. 

### Browser fixtures were overhauled

The webdriver-recorder 4.0 changes rendered some of the fixtures needed here obsolete, but this test suite also has several edge cases with regards to cookies that can only be handled by creating what Glen coined "fresh" instances. To ensure these fresh instances are created safely with respect to their current environment, I added two fixtures that handle all the complexity of setting them up:

- `fresh_browser` will configure a browser instance, connect it to a running chromedriver service, and clean up after your test is finished.
- `fresh_class_browser` can be applied to class-based tests to automatically set `self.browser` to a "fresh" instance whose lifetime is scoped to the class; as with `fresh_browser`, this instance cleans up after itself when all tests in the class have finished.

### Updates the `login_submit_form` fixture:

- Renamed this fixture to `log_in_netid` so that the flow is descriptive of what the browser is doing, rather than describing a DOM element. 
- The `password` argument is now optional, and will default to the secret (i.e., correct) password.
- Automatically detects whether the entered password is the correct one, and asserts the success case if so. 
- Use `assert_success=False` if you know that the login won't succeed
- Use `select_iframe=False` if you know the iframe context is already selected
- Use `match_service_provider=sp` when you expect the login to succeed (otherwise, only the success message, not the idp domain, will be matched).

### Updates the `two_fa_submit_form` fixture:

- Renamed this fixture to `enter_duo_passcode` so that the flow is descriptive of what the browser is doing, and also to remove the naming issue introduced by the number 2.
- The `passcode` argument is now optional, and will default to the secret (i.e., correct) passcode.
- Automatically detects whether the entered passcode is the correct one, and asserts either the success state or the invalid passcode state. You can toggle both checks using `assert_success=True|False` and `assert_failure=True|False`, respectively.
- Use `match_service_provider=sp` when you expect the login to succeed (otherwise, only the success message, not the idp domain, will be matched).

### Adds the fixture `sp_shib_url`

Since almost every test case used the pattern `f"{sp_url(sp)/shib{test_env}"`, I refactored this into a fixture and replaced all instances.

Use it like so:

```python
# assuming test_env == 'prod' and sp == ServiceProviderInstance.diafine6

sp_shib_url(sp)   # https://diafine6.sandbox.iam.s.uw.edu/shibprod
sp_shib_url(sp, append='mfa')  # https://diafine6.sandbox.iam.s.uw.edu/shibprodmfa
```

This accounts for the current `test_env`.

### CODE REVIEW

**Comment by tomthorogood** [comment]
I'll handle this merge conflict and update the PR sometime today

**Comment by tomthorogood** [comment]
Updated to fix merge conflicts

File: tests/access_control/test_auto_2fa.py [hunk 1]
```diff
@@ -1,118 +1,95 @@
...
+            self.log_in_netid(self.browser, netid3, match_service_provider=sp)
 
         sp = ServiceProviderInstance.diafine7
         with self.utils.using_test_sp(sp):
-            fresh_browser.get(f'{self.sp_url(sp)}/shib{self.test_env}')
-            self.two_fa_submit_form(fresh_browser)
-            fresh_browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
+            self.browser.get(self.sp_shib_url(sp))
+            self.enter_duo_passcode(self.browser)
+            self.browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
```
Comment (goulter) [inline tests/access_control/test_auto_2fa.py]: Looks like line 31 would be redundant based on the definition of `self.enter_duo_passcode`

File: tests/access_control/test_auto_2fa.py [hunk 2]
```diff
@@ -1,118 +1,95 @@
...
-        #
-        fresh_browser = Chrome()
+        # Prompted for pwd then 2FA on diafine7.
         sp = ServiceProviderInstance.diafine6
         with self.utils.using_test_sp(sp):
-            fresh_browser.get(f'{self.sp_url(sp)}/shib{self.test_env}')
-            self.login_submit_form(fresh_browser, netid3, self.password)
-            fresh_browser.wait_for_tag('h2', f'{self.sp_domain(sp)} sign-in success!')
+            with self.browser.tab_context():
+                self.browser.get(self.sp_shib_url(sp))
```
Comment (goulter) [inline tests/access_control/test_auto_2fa.py:40]: what is `self.browser.tab_context():`?

File: tests/access_control/test_auto_2fa.py [hunk 3]
```diff
@@ -1,118 +1,95 @@
...
+                self.browser.get(self.sp_shib_url(sp, append='force'))
+                self.log_in_netid(self.browser, netid3, match_service_provider=sp, assert_success=False)
+                self.enter_duo_passcode(self.browser)
 
     def test_auto_2fa_d(self, netid3):
         """
-        AC-1 Part D. No prompts on diafine7.
+        AC-1 Part D
         """
-        fresh_browser = Chrome()
```
Comment (goulter) [inline tests/access_control/test_auto_2fa.py:73]: What is the replacement for the new browser instance that was getting created on line 73?

File: tests/test_attributes.py [hunk 4]
```diff
@@ -57,36 +54,17 @@ def test_attributes(browser, netid, utils, sp_url, sp_domain, secrets, test_env)
...
-                    'student@washington.edu',
-                    'staff@washington.edu',
-                    'member@washington.edu',
-                }, f'Unexpected attributes found in comparison with: {found_values}'
-            else:
-                # matches on exact string only. mail won't count as a match for uwEduEmail and uwEduEmail won't match
-                # mail
-                assert f'{key} = {value}' in attribute_data
-
-        fresh_browser.close()
```
Comment (goulter) [inline tests/test_attributes.py:92]: Lines 73-78 and 80-86 should not be removed, they account for testing `eduPersonScopedAffiliation`.

File: tests/test_2fa_duo.py [hunk 5]
```diff
@@ -3,264 +3,205 @@
...
+        utils, sp_shib_url, sp_url, log_in_netid,
+        sp_domain, secrets, netid3, test_env, fresh_browser, enter_duo_passcode):
     """
     2FA-9 Remember me cookie
     """
-    # chrome_options = webdriver.ChromeOptions()
-    # chrome_options.add_experimental_option("prefs", {"profile.default_content_settings.cookies": 1})
-    # fresh_browser = Chrome(options=chrome_options)
-    fresh_browser = Chrome()
+    fresh_browser
```
Comment (goulter) [inline tests/test_2fa_duo.py]: Does `fresh_browser` have a value?

**Review by goulter — COMMENTED** [review]
A few questions, one of which may spawn a change request :)

File: tests/access_control/test_auto_2fa.py (hunk 1 above)
Comment (tomthorogood) [reply tests/access_control/test_auto_2fa.py]: Correct, thanks!

File: tests/access_control/test_auto_2fa.py (hunk 2 above)
Comment (tomthorogood) [reply tests/access_control/test_auto_2fa.py:40]: Simply opens a tab, then closes it at the end of the block. Only used here because you had previously explicitly opened and closed a tab in this particular test.

https://github.com/UWIT-IAM/webdriver-recorder/blob/master/webdriver_recorder/browser.py#L264-L270

File: tests/test_attributes.py (hunk 4 above)
Comment (tomthorogood) [reply tests/test_attributes.py:92]: I feel like this might have been a merge conflict problem. I'll restore the intent here. I wouldn't have purposely deleted that stuff.

File: tests/access_control/test_auto_2fa.py (hunk 3 above)
Comment (tomthorogood) [reply tests/access_control/test_auto_2fa.py:73]: `self.browser` is always a `fresh_browser` for every test in the `access_control` package, if you look at the `AccessControlTestBase` class.

File: tests/test_2fa_duo.py (hunk 5 above)
Comment (tomthorogood) [reply tests/test_2fa_duo.py]: of course it has a value, it's a fixture, but no, this line shouldn't be here.

File: tests/test_attributes.py (hunk 4 above)
Comment (tomthorogood) [reply tests/test_attributes.py:92]: I think these weren't so much "removed" as "accidentally merge-conflicted away" --- i'll make sure to add those test cases back

File: tests/access_control/test_auto_2fa.py (hunk 2 above)
Comment (tomthorogood) [reply tests/access_control/test_auto_2fa.py:40]: (Not sure if the tab is actually necessary, but preserved the flow you wrote)

File: tests/access_control/test_auto_2fa.py (hunk 3 above)
Comment (goulter) [reply tests/access_control/test_auto_2fa.py:73]: I see it now, cool.

File: tests/test_attributes.py (hunk 4 above)
Comment (goulter) [reply tests/test_attributes.py:92]: Thx

**Review by goulter — APPROVED** [review]
